src/clj/rems/api/applications_v2.clj

```diff
@@ -0,0 +1,810 @@
+                                            :resource/ext-id (:resource/ext-id resource)})
+                                         (:application/resources event))
+             :application/licenses (map (fn [license]
+                                          {:license/id (:license/id license)
+                                           :license/accepted false})
+                                        (:application/licenses event))
+             :application/events []
+             :application/form {:form/id (:form/id event)
+                                :form/fields []}
+             :application/workflow {:workflow/id (:workflow/id event)
+                                    :workflow/type (:workflow/type event)
+                                    ;; TODO: or would :workflow.dynamic/state be more appropriate?
+                                    :workflow/state :rems.workflow.dynamic/draft ; TODO: other workflows
+                                    :workflow.dynamic/handlers (:workflow.dynamic/handlers event)})
+      (set-role-permissions {:applicant #{::dynamic/save-draft
+                                          ::dynamic/submit}})))
+
+(defn- set-accepted-licences [licenses accepted-licenses]
+  (map (fn [license]
+         (assoc license :license/accepted (contains? accepted-licenses (:license/id license))))
+       licenses))
+
+(defmethod event-type-specific-application-view :application.event/draft-saved
+  [application event]
+  (-> application
+      (assoc :application/modified (:event/time event))
+      (assoc-in [:application/form :form/fields] (map (fn [[field-id value]]
+                                                        {:field/id field-id
+                                                         :field/value value})
+                                                      (:application/field-values event)))
+      (update :application/licenses set-accepted-licences (:application/accepted-licenses event))))
+
+(defmethod event-type-specific-application-view :application.event/member-invited
+  [application event]
+  application)
+
+(defmethod event-type-specific-application-view :application.event/member-added
+  [application event]
+  application)
+
+(defmethod event-type-specific-application-view :application.event/submitted
+  [application event]
+  (-> application
+      (assoc :workflow/state ::dynamic/submitted)
+      (set-role-permissions {:applicant #{}
+                             :handler #{::dynamic/approve
+                                        ::dynamic/reject
+                                        ::dynamic/return
+                                        ::dynamic/request-decision
+                                        ::dynamic/request-comment}})))
+
+(defmethod event-type-specific-application-view :application.event/returned
+  [application event]
+  application)
+
+(defmethod event-type-specific-application-view :application.event/comment-requested
```

Comment:
Would this be the place where you add the commenter permissions? Isn't there significant overlap with possible commands still?

Reply:
Yes, that would be the place. The idea is to derive the possible commands from the permissions.